pysb/core.py

```diff
@@ -1643,6 +1647,14 @@ def __getitem__(self, key):
         else:
             return self._map[key]
 
+    def __getattr__(self, name):
+        if name == '_map':
```

Comment:
What is this guarding against?

Reply:
Infinite recursion. Without this check, `self._map[key]` in `__getitem__` calls `__getattr__`, which calls `__getitem__`.....

Reply:
Did you actually witness the recursion? I'm pretty sure `__getattr__` is only called for attributes that don't actually exist in the object. Maybe you were thinking of `__getattribute__` which is called unconditionally?

Reply:
Yes, I did witness it. Commenting out the `if` clause gives the following:

```
$ nosetests pysb.tests.test_core
.........E.................
======================================================================
ERROR: pysb.tests.test_core.test_model_pickle
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/.virtualenvs/pysb/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/alex/git/VU/pysb/pysb/testing/__init__.py", line 18, in inner
    func(*args, **kwargs)
  File "/Users/alex/git/VU/pysb/pysb/tests/test_core.py", line 112, in test_model_pickle
    model2 = pickle.loads(pickle.dumps(model))
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 1388, in loads
    return Unpickler(file).load()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 1221, in load_build
    setstate = getattr(inst, "__setstate__", None)
  File "/Users/alex/git/VU/pysb/pysb/core.py", line 1839, in __getattr__
    return self[name]
  File "/Users/alex/git/VU/pysb/pysb/core.py", line 1833, in __getitem__
    return self._map[key]
  File "/Users/alex/git/VU/pysb/pysb/core.py", line 1839, in __getattr__
    return self[name]
  File "/Users/alex/git/VU/pysb/pysb/core.py", line 1833, in __getitem__
    return self._map[key]
  File "/Users/alex/git/VU/pysb/pysb/core.py", line 1839, in __getattr__
    return self[name]
  File "/Users/alex/git/VU/pysb/pysb/core.py", line 1833, in __getitem__
    return self._map[key]
...<snip>...
RuntimeError: maximum recursion depth exceeded
```

Reply:
Oh, that looks like a pickling-specific problem. A little googling for `__getattr__` and pickling led to this bug report and the resulting section added to the docs:

https://bugs.python.org/issue5370
https://docs.python.org/2.7/library/pickle.html#pickling-and-unpickling-normal-class-instances

> At unpickling time, some methods like `__getattr__()`, `__getattribute__()`, or `__setattr__()` may be called upon the instance. In case those methods rely on some internal invariant being true, the type should implement either `__getinitargs__()` or `__getnewargs__()` to establish such an invariant; otherwise, neither `__new__()` nor `__init__()` will be called. 

Can we fix this at the level of the pickle protocol? The current solution is confusing to me.